##### Here I am implementing a classification model for Telecom data.

In [1]:
# imports 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd

In [2]:
# reading dataset into a pandas dataframe
df=pd.read_csv('Churn-data.csv')

In [3]:
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [4]:
df.shape

(7043, 21)

In [5]:
df.head().T

0             1               2  \
customerID              7590-VHVEG    5575-GNVDE      3668-QPYBK   
gender                      Female          Male            Male   
SeniorCitizen                    0             0               0   
Partner                        Yes            No              No   
Dependents                      No            No              No   
tenure                           1            34               2   
PhoneService                    No           Yes             Yes   
MultipleLines     No phone service            No              No   
InternetService                DSL           DSL             DSL   
OnlineSecurity                  No           Yes             Yes   
OnlineBackup                   Yes            No             Yes   
DeviceProtection                No           Yes              No   
TechSupport                     No            No              No   
StreamingTV                     No            No              No   
StreamingMovies                 No            No              No   
Contract            Month-to-month      One year  Month-to-month   
PaperlessBilling               Yes            No             Yes   
PaymentMethod     Electronic check  Mailed check    Mailed check   
MonthlyCharges               29.85         56.95           53.85   
TotalCharges                 29.85        1889.5          108.15   
Churn                           No            No             Yes   

                                          3                 4  
customerID                       7795-CFOCW        9237-HQITU  
gender                                 Male            Female  
SeniorCitizen                             0                 0  
Partner                                  No                No  
Dependents                               No                No  
tenure                                   45                 2  
PhoneService                             No               Yes  
MultipleLines              No phone service                No  
InternetService                         DSL       Fiber optic  
OnlineSecurity                          Yes                No  
OnlineBackup                             No                No  
DeviceProtection                        Yes                No  
TechSupport                             Yes                No  
StreamingTV                              No                No  
StreamingMovies                          No                No  
Contract                           One year    Month-to-month  
PaperlessBilling                         No               Yes  
PaymentMethod     Bank transfer (automatic)  Electronic check  
MonthlyCharges                         42.3              70.7  
TotalCharges                        1840.75            151.65  
Churn                                    No               Yes

In [6]:
# checking datatypes
df.dtypes

customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                object
dtype: object

In [7]:
# Here the total charges column is shown as object type(string) so converting that to numeric
df['TotalCharges']=pd.to_numeric(df['TotalCharges'],errors='coerce')
df['TotalCharges']=df['TotalCharges'].fillna(0)

In [8]:
# Converting all the column names to lower case and the columns having string values also to lower case and replace space with _

In [9]:
df.columns=df.columns.str.lower().str.replace(' ','_')
string_cols=list(df.dtypes[df.dtypes==object].index)

In [10]:
for i in string_cols:
    df[i]=df[i].str.lower().str.replace(' ','_')

In [11]:
# Converting the datatype of Churn vairable to integer
df.churn=(df.churn=='yes').astype(int)

In [12]:
# Checking the datatypes now
df.head().T

0             1               2  \
customerid              7590-vhveg    5575-gnvde      3668-qpybk   
gender                      female          male            male   
seniorcitizen                    0             0               0   
partner                        yes            no              no   
dependents                      no            no              no   
tenure                           1            34               2   
phoneservice                    no           yes             yes   
multiplelines     no_phone_service            no              no   
internetservice                dsl           dsl             dsl   
onlinesecurity                  no           yes             yes   
onlinebackup                   yes            no             yes   
deviceprotection                no           yes              no   
techsupport                     no            no              no   
streamingtv                     no            no              no   
streamingmovies                 no            no              no   
contract            month-to-month      one_year  month-to-month   
paperlessbilling               yes            no             yes   
paymentmethod     electronic_check  mailed_check    mailed_check   
monthlycharges               29.85         56.95           53.85   
totalcharges                 29.85        1889.5          108.15   
churn                            0             0               1   

                                          3                 4  
customerid                       7795-cfocw        9237-hqitu  
gender                                 male            female  
seniorcitizen                             0                 0  
partner                                  no                no  
dependents                               no                no  
tenure                                   45                 2  
phoneservice                             no               yes  
multiplelines              no_phone_service                no  
internetservice                         dsl       fiber_optic  
onlinesecurity                          yes                no  
onlinebackup                             no                no  
deviceprotection                        yes                no  
techsupport                             yes                no  
streamingtv                              no                no  
streamingmovies                          no                no  
contract                           one_year    month-to-month  
paperlessbilling                         no               yes  
paymentmethod     bank_transfer_(automatic)  electronic_check  
monthlycharges                         42.3              70.7  
totalcharges                        1840.75            151.65  
churn                                     0                 1

##### Dividing dataset into 3 sets: train,test and validation .I am using sklearn train_test_split fucction with a fixed random state by first using this on whole dataset and then again using on obtained dataset.

In [13]:
from sklearn.model_selection import train_test_split

In [14]:
X_T,y_test=train_test_split(df,test_size=0.20,random_state=1)
X_train,X_val=train_test_split(X_T,test_size=0.33,random_state=10)
y_train=X_train.churn.values
y_val=X_val.churn.values

In [15]:
# Removing churn variable from training dataset.
del X_train['churn']
del X_val['churn']

##### EDA

In [16]:
# Checking null values
X_T.isnull().sum()

customerid          0
gender              0
seniorcitizen       0
partner             0
dependents          0
tenure              0
phoneservice        0
multiplelines       0
internetservice     0
onlinesecurity      0
onlinebackup        0
deviceprotection    0
techsupport         0
streamingtv         0
streamingmovies     0
contract            0
paperlessbilling    0
paymentmethod       0
monthlycharges      0
totalcharges        0
churn               0
dtype: int64

In [17]:
# There are no null values thats a sign of releif.
# So now checking the distribution of churn
X_T.churn.value_counts()

0    4113
1    1521
Name: churn, dtype: int64

In [18]:
# There is a huge difference in 2 classes so its a case of imbalanced classification. Calculating churn rate as follows:
global_mean=X_T.churn.mean()
round(global_mean,2)

0.27

In [19]:
# As we are habing categorical as well as numerical values we need to separate them out for better analysis.
categorical = ['gender', 'seniorcitizen', 'partner', 'dependents','phoneservice', 'multiplelines', 'internetservice',
'onlinesecurity', 'onlinebackup', 'deviceprotection','techsupport', 'streamingtv', 'streamingmovies','contract', 'paperlessbilling',
 'paymentmethod']
numerical = ['tenure', 'monthlycharges', 'totalcharges']

In [20]:
# Checking the number of unique values in categorical features
X_T[categorical].nunique()

gender              2
seniorcitizen       2
partner             2
dependents          2
phoneservice        2
multiplelines       3
internetservice     3
onlinesecurity      3
onlinebackup        3
deviceprotection    3
techsupport         3
streamingtv         3
streamingmovies     3
contract            3
paperlessbilling    2
paymentmethod       4
dtype: int64

##### Feature Selection
##### Firstly checking which categorical features are important

##### Churn Rate : Looking for churn rate for each variable.I'll calculate the group churn rate and compare it to global churn rate. The group here is the collection of values of a particular type in a variable. If the difference between the rates is small, the value is not important when predicting churn because this group of customers is not really different from the rest of the customers. On the other hand, if the difference is not small, I will consider that group and variable.

##### For example: Taking gender variable:

In [21]:
female_mean=X_T[X_T.gender == 'female'].churn.mean()
print('gender==female:',round(female_mean, 2))
male_mean=X_T[X_T.gender == 'male'].churn.mean()
print('gender==male:',round(male_mean, 2))

gender==female: 0.28
gender==male: 0.26


##### The difference between the means of males and females is quite small so it will not to a good option to consider gender as a variable while making predictions.
##### In addition to looking at the difference between the group rate and the global rate,I'll look at the ratio between them. This ration is known as risk ratio , where risk refers to the risk of having the effect of churning

In [22]:
# Implementing thr risk ratio for all categorical variables:
from IPython.display import display
for i in categorical:
    df_=X_T.groupby(by=i).churn.agg(['mean'])
    df_['difference']=df_['mean']-global_mean
    df_['risk_ratio']=df_['mean']/global_mean
    display(df_)

mean  difference  risk_ratio
gender                                  
female  0.276824    0.006856    1.025396
male    0.263214   -0.006755    0.974980

mean  difference  risk_ratio
seniorcitizen                                  
0              0.242270   -0.027698    0.897403
1              0.413377    0.143409    1.531208

mean  difference  risk_ratio
partner                                  
no       0.329809    0.059841    1.221659
yes      0.205033   -0.064935    0.759472

mean  difference  risk_ratio
dependents                                  
no          0.313760    0.043792    1.162212
yes         0.165666   -0.104302    0.613651

mean  difference  risk_ratio
phoneservice                                  
no            0.241316   -0.028652    0.893870
yes           0.273049    0.003081    1.011412

mean  difference  risk_ratio
multiplelines                                     
no                0.257407   -0.012561    0.953474
no_phone_service  0.241316   -0.028652    0.893870
yes               0.290742    0.020773    1.076948

mean  difference  risk_ratio
internetservice                                  
dsl              0.192347   -0.077621    0.712482
fiber_optic      0.425171    0.155203    1.574895
no               0.077805   -0.192163    0.288201

mean  difference  risk_ratio
onlinesecurity                                       
no                   0.420921    0.150953    1.559152
no_internet_service  0.077805   -0.192163    0.288201
yes                  0.153226   -0.116742    0.567570

mean  difference  risk_ratio
onlinebackup                                         
no                   0.404323    0.134355    1.497672
no_internet_service  0.077805   -0.192163    0.288201
yes                  0.217232   -0.052736    0.804660

mean  difference  risk_ratio
deviceprotection                                     
no                   0.395875    0.125907    1.466379
no_internet_service  0.077805   -0.192163    0.288201
yes                  0.230412   -0.039556    0.853480

mean  difference  risk_ratio
techsupport                                          
no                   0.418914    0.148946    1.551717
no_internet_service  0.077805   -0.192163    0.288201
yes                  0.159926   -0.110042    0.592390

mean  difference  risk_ratio
streamingtv                                          
no                   0.342832    0.072864    1.269897
no_internet_service  0.077805   -0.192163    0.288201
yes                  0.302723    0.032755    1.121328

mean  difference  risk_ratio
streamingmovies                                      
no                   0.338906    0.068938    1.255358
no_internet_service  0.077805   -0.192163    0.288201
yes                  0.307273    0.037305    1.138182

mean  difference  risk_ratio
contract                                        
month-to-month  0.431701    0.161733    1.599082
one_year        0.120573   -0.149395    0.446621
two_year        0.028274   -0.241694    0.104730

mean  difference  risk_ratio
paperlessbilling                                  
no                0.172071   -0.097897    0.637375
yes               0.338151    0.068183    1.252560

mean  difference  risk_ratio
paymentmethod                                              
bank_transfer_(automatic)  0.168171   -0.101797    0.622928
credit_card_(automatic)    0.164339   -0.105630    0.608733
electronic_check           0.455890    0.185922    1.688682
mailed_check               0.193870   -0.076098    0.718121

##### By looking at the differences and the risks, I can identify the most informative features: the features that are helpful for detecting churn.it’s hard to use them to say what the most important feature is so I am using mutual information metric from sklearn which can measure the degree of dependency between a categorical variable and the target variable.if the mutual information between a categorical variable and the target is high, this categorical variable is useful for predicting the target. On the other hand, if the mutual information is low, the categorical variable and the target are independent,so the variable will not be useful for predictions.

In [23]:
from sklearn.metrics import mutual_info_score
def calculate_mut_info(a):
    return mutual_info_score(a,X_T.churn)
df_mi=X_T[categorical].apply(calculate_mut_info)
df_mi=df_mi.sort_values().to_frame(name='Mutual Info')
display(df_mi)

Mutual Info
gender               0.000117
phoneservice         0.000229
multiplelines        0.000857
seniorcitizen        0.009410
partner              0.009968
dependents           0.012346
paperlessbilling     0.017589
streamingmovies      0.031581
streamingtv          0.031853
paymentmethod        0.043210
deviceprotection     0.043453
onlinebackup         0.046923
internetservice      0.055868
techsupport          0.061032
onlinesecurity       0.063085
contract             0.098320

##### contract, onlinesecurity, and techsupport are among the most important features.Gender and phone service are least important so I will not use them for training model.

##### Feature Selection for numerical variables:

##### Mutual info can't be used between numerical variable and binary target variable so using Pearson correlation coefficient.

In [24]:
X_T[numerical].corrwith(X_T.churn).to_frame('P_correlation')

P_correlation
tenure              -0.351885
monthlycharges       0.196805
totalcharges        -0.196353

In [25]:
X_T.groupby(by='churn')[numerical].mean()

tenure  monthlycharges  totalcharges
churn                                         
0      37.531972       61.176477   2548.021627
1      18.070348       74.521203   1545.689415

##### Feature Engineering: The features which I will use need to be converted to numbers as Machine learning algorithms don't understand any other types of values so I will use One Hot encoding to convert categorical values to numerical and I am using DictVectoizer for this.

In [27]:
# importing and making an object
from sklearn.feature_extraction import DictVectorizer
dv=DictVectorizer(sparse=False)

In [30]:
train_dic=X_train[numerical+categorical].to_dict(orient='records')
dv.fit(train_dic)

DictVectorizer(sparse=False)

##### The to_dict has an important parameter orient that can has multiple values but I chose records here which is list like and has format as : {['column'->value],['column'->value],['column'->value].....}

In [31]:
X_train_dict=dv.transform(train_dic)

In [33]:
X_train_dict.shape

(3774, 45)

##### Logistic Regression Model Training:

In [35]:
# importing from sklearn
from sklearn.linear_model import LogisticRegression

In [36]:
model=LogisticRegression()
model.fit(X_train_dict,y_train)

C:\Users\aksha\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [37]:
# Measure performance using validation set
val_dict=X_val[categorical + numerical].to_dict(orient='records')
X_val_dict=dv.transform(val_dict)

In [38]:
model.predict_proba(X_val_dict)

array([[0.98209632, 0.01790368],
       [0.97494926, 0.02505074],
       [0.35715346, 0.64284654],
       ...,
       [0.95307822, 0.04692178],
       [0.76631927, 0.23368073],
       [0.88231667, 0.11768333]])

##### The first column above shows that there is no churn and second shows there is churn so selecting second column only.

In [39]:
y_pred=model.predict_proba(X_val_dict)[:, 1]
y_pred

array([0.01790368, 0.02505074, 0.64284654, ..., 0.04692178, 0.23368073,
       0.11768333])

##### These are the soft predictions sthat a customer will churn so I will convert them into actual true and false predictions by using a thershold value of 0.5 which is generally taken.

In [40]:
churn=y_pred>0.5

I am using accuracy for measuring model performance : calculate the number of times the predictions match the actual value, I will use it for measuring the quality of the model.This quality measure is called accuracy.

In [41]:
(y_val==churn).mean()

0.7876344086021505

##### Although I used accuracy as a measure to evaluate the performance of model there are many more metrics also and 78.76% accuracy is good without performing much things.